cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,304 @@
+      : _comparator(t, t, true), _map{map} {}
+
+  /**
+   * @brief Returns true if the two rows at the specified indices in the permuted
+   * order are equivalent.
+   *
+   * For example, comparing rows `i` and `j` is
+   * equivalent to comparing rows `map[i]` and `map[j]` in the original table.
+   *
+   * @param lhs The index of the first row
+   * @param rhs The index of the second row
+   * @returns if the two specified rows in the permuted order are equivalent
+   */
+  CUDA_DEVICE_CALLABLE
+  bool operator()(gdf_size_type lhs, gdf_size_type rhs) {
+    return _comparator(_map[lhs], _map[rhs]);
+  }
+};
+
+} // namespace anonymous
+
+
+namespace cudf {
+
+namespace detail {
+
+gdf_column const& groupby::key_sort_order() {
+  if (_key_sorted_order)
+    return *_key_sorted_order;
+
+  _key_sorted_order = gdf_col_pointer(
+    new gdf_column(
+      allocate_column(gdf_dtype_of<gdf_index_type>(),
+                      _keys.num_rows(),
+                      false,
+                      gdf_dtype_extra_info{},
+                      _stream)),
+    [](gdf_column* col) { gdf_column_free(col); });
+
+  if (_include_nulls ||
+      !cudf::has_nulls(_keys)) {  // SQL style
+    gdf_context context{};
+    context.flag_groupby_include_nulls = true;
+    CUDF_TRY(gdf_order_by(_keys.begin(), nullptr,
+                          _keys.num_columns(), _key_sorted_order.get(),
+                          &context));
+  } else {  // Pandas style
+    auto keys_row_bitmask = row_bitmask(_keys, _stream);
+
+   // Temporarily replace the first column's bitmask with one that indicates the 
+   // presence of a null value within a row.  This allows moving all rows that contain
+   // a null value to the end of the sorted order. 
+    gdf_column null_row_representative = *(_keys.get_column(0));
+    null_row_representative.valid =
+        reinterpret_cast<gdf_valid_type*>(keys_row_bitmask.data().get());
+
+    cudf::table keys{_keys};
+    std::vector<gdf_column*> modified_keys(keys.begin(), keys.end());
+    modified_keys[0] = &null_row_representative;
+    cudf::table modified_keys_table(modified_keys.data(),
+                                    modified_keys.size());
+
+    gdf_context temp_ctx;
+    temp_ctx.flag_null_sort_behavior = GDF_NULL_AS_LARGEST;
+
+    CUDF_TRY(gdf_order_by(modified_keys_table.begin(), nullptr,
+                          modified_keys_table.num_columns(),
+                          _key_sorted_order.get(), &temp_ctx));
+
+    // All rows with one or more null values are at the end of the resulting sorted order.
+    // The number of rows w/o null values `n` is indicated by number of valid bits
+    // in the row bitmask. When `include_nulls == false`, then only rows `[0, n)` 
+    // in the sorted order are considered for grouping. 
+    CUDF_TRY(gdf_count_nonzero_mask(
+        reinterpret_cast<gdf_valid_type*>(keys_row_bitmask.data().get()),
+        _keys.num_rows(),
+        &_num_keys));
```

Comment:
All of this can be solved by making `_num_keys` memoized. It depends only on `key_sort_order` being computed first. So I'll just do that. Something like this:
```c++
gdf_size_type num_keys() { 
  // sort order needs to be computed first
  key_sort_order();
  return _num_keys;
}
```
`num_keys` **are** the `num_effective_keys` and `keys.num_rows()` does not need to be a part of this groupby object. For all user intents and purposes, once it's constructed with `include_nulls=false`, `num_keys` **is** the number of keys there are in this groupby object.

The documentation already mentions what `num_keys()` gives.

Reply:
> num_keys is the number of keys there are in this groupby object.

I would agree with you if not for `unsorted_keys_labels`, which can return a potentially different number of elements than `num_keys()`. 

The issue is consistency.

If all APIs only interact with the "effective" keys, then I agree there's no need to differentiate between `num_keys()` and `num_effective_keys()`. 

However, there are currently a mix of APIs that refer to either the "effective" number of keys, or the original number of keys, and there needs to be a way to meaningfully differentiate between those two things.

Reply:
> All of this can be solved by making _num_keys memoized. It depends only on key_sort_order being computed first.

All internal references to `_num_keys` should then be replaced by calls to `num_keys()`.

Reply:
Actually, I should make the `unsorted_keys_labels()` private again. It's used only for sorting and grouping which is publicly handled by `sort_values()`.
The problem is, if the user needs to be aware of the distinction between the number of keys before and after, and how it's done, then it's not a good abstraction.

Reply:
> All internal references to `_num_keys` should then be replaced by calls to `num_keys()`

They already are.

Reply:
> Actually, I should make the unsorted_keys_labels() private again. It's used only for sorting and grouping which is publicly handled by sort_values().

This is a good solution. `unsorted_keys_labels()` is really the source of my consternation. 

> The problem is, if the user needs to be aware of the distinction between the number of keys before and after, and how it's done, then it's not a good abstraction.

Yes, I think that is why I've struggled with the `num_keys()` vs `num_effective_keys()` and therefore been a PITA :)

Reply:
```
gdf_size_type num_keys() { 
  // sort order needs to be computed first
  key_sort_order();
  return _num_keys;
}
```

While this solution for memoizing `num_keys` will work, it's a bitter heavier handed than what is necessary. In reality, all you need to do is compute the row bitmask for `keys` and the compute the null count. You could memoize both the row bitmask and the null count for usage in `key_sort_order()`.

Reply:
> Actually, I should make the `unsorted_keys_labels()` private again

I have a bad feeling that I'll have to eat my words when I come across an agg I have to write that **does** have to use this 😅

Reply:
>  In reality, all you need to do is compute the row bitmask for keys and the compute the null count. 

Oh, and a shortcut to avoid computing the row bitmask is to first check if `cudf::has_nulls(_keys)`.

Reply:
Do you have a suggestion for checking if a scalar is initialized?

Reply:
> Do you have a suggestion for checking if a scalar is initialized?

You could default it to `-1`, and check if its positive.